lib/grakn/channel.ex

```diff
@@ -14,27 +14,39 @@ defmodule Grakn.Channel do
   end
 
   @spec open_transaction(t(), Transaction.request()) ::
-          {:ok, Grakn.Transaction.t(), String.t()} | {:error, any()}
+          {:ok, t(), Grakn.Transaction.t(), String.t()} | {:error, any()}
   def open_transaction(channel, %Transaction{type: type, opts: opts} = tx_request) do
-    with {:ok, {session_id, cached?}} <- fetch_or_open_session(channel, tx_request) do
-      with {:ok, tx} <- Grakn.Transaction.new(channel, opts),
-           {:ok, tx} <- Transaction.open(tx, session_id, type, opts) do
-        {:ok, tx, session_id}
-      else
-        error ->
-          may_retry_session(channel, tx_request, error, cached?)
-      end
+    case fetch_or_open_session(channel, tx_request) do
+      {:ok, {session_id, cached?}} ->
+        with {:ok, tx} <- Grakn.Transaction.new(channel, opts),
+             {:ok, tx} <- Transaction.open(tx, session_id, type, opts) do
+          {:ok, channel, tx, session_id}
+        else
+          error ->
+            may_retry(channel, tx_request, error, cached?)
+        end
+
+      error ->
+        may_retry(channel, tx_request, error, false)
```

Comment:
We could add some tests for this new logic, what do you think?

Reply:
The tests would be hacky, as I need to add API, which expose internals of channel implementation to simulate observed behaviour in production.

What do you think, should we do this?